compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -121,44 +123,92 @@ std::optional<Value> hoistOneStaticallyBoundAllocation(
+        return OpFoldResult(builder.getIndexAttr(constantBound));
+      }
+
+      if (!vscale)
+        vscale = builder.create<vector::VectorScaleOp>(loc);
+      return affine::materializeComputedBound(
+          builder, loc, ub->map, {std::make_pair(vscale, std::nullopt)});
+    }
+    // Non-scalable target: Assume everything is fixed-size.
+    auto ub = ValueBoundsConstraintSet::computeConstantBound(
+        presburger::BoundType::UB, value, {}, /*stopCondition=*/nullptr,
+        /*closedUB=*/true);
+    if (failed(ub))
+      return failure();
+
+    return OpFoldResult(builder.getIndexAttr(*ub));
+  };
+
   /// For the dynamic but bounded case, insert an allocation of the shape of the
   /// bounds, and a subview of the required size to be used as a replacement.
-  SmallVector<int64_t> staticShape;
+
+  SmallVector<OpFoldResult> allocSizes;
   SmallVector<OpFoldResult> subviewSizes;
-  staticShape.reserve(allocLikeType.getRank());
+  allocSizes.reserve(allocLikeType.getRank());
   subviewSizes.reserve(allocLikeType.getRank());
 
-  int index = 0;
-  for (auto dimSize : allocLikeType.getShape()) {
-    if (!ShapedType::isDynamic(dimSize)) {
-      staticShape.push_back(dimSize);
-      subviewSizes.push_back(builder.getIndexAttr(dimSize));
-      continue;
-    }
-    Value dynamicSize = dynamicSizes[index++];
-    auto ub = ValueBoundsConstraintSet::computeConstantBound(
-        presburger::BoundType::UB, dynamicSize, /*dim=*/std::nullopt,
-        /*stopCondition=*/nullptr, /*closedUB=*/true);
-    if (failed(ub)) {
-      return std::nullopt;
-    }
-    staticShape.push_back(ub.value());
-    subviewSizes.push_back(dynamicSize);
-  }
-  SmallVector<OpFoldResult> offsets(allocLikeType.getRank(),
-                                    builder.getIndexAttr(0));
-  SmallVector<OpFoldResult> strides(allocLikeType.getRank(),
-                                    builder.getIndexAttr(1));
-
   Value allocation;
   {
     OpBuilder::InsertionGuard g(builder);
     builder.setInsertionPointToStart(&funcOp.getFunctionBody().front());
+
+    int index = 0;
+    for (auto dimSize : allocLikeType.getShape()) {
+      if (!ShapedType::isDynamic(dimSize)) {
+        auto dimSizeAttr = builder.getIndexAttr(dimSize);
+        allocSizes.push_back(dimSizeAttr);
+        subviewSizes.push_back(dimSizeAttr);
+        continue;
+      }
+
+      Value dynamicSize = dynamicSizes[index++];
+      auto ub = computeAllocationBound(dynamicSize);
+      if (failed(ub))
+        return std::nullopt;
+
+      allocSizes.push_back(*ub);
+      subviewSizes.push_back(dynamicSize);
+    }
+
+    // FIXME: The `AllocLikeOp::build()` method for OpFoldResult drops the
+    // layout, so we have to resolve the static/dynamic values here.
+    SmallVector<int64_t> staticShape;
+    SmallVector<Value> dynamicSizes;
+    staticShape.reserve(allocLikeType.getRank());
+    dynamicSizes.reserve(allocLikeType.getRank());
```

Comment:
[optional] I'd avoid using the trick here. This kind of tricks should be happening in `dispatchIndexOpFoldResults` implementation details.